File: 653-two-sum-iv-input-is-a-bst/653-two-sum-iv-input-is-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {boolean}
 */
var findTarget = function(root, k) {
    const cache = new Set();
    
    let level = [root];
    
    while (level.length > 0) {
        const nextLevel = [];
        
        for (const node of level) {
            if (cache.has(node.val)) {
                return true;
            }
            cache.add(k - node.val);
            
            node.left && nextLevel.push(node.left);
            node.right && nextLevel.push(node.right);
        }
        level = nextLevel;
    }
    return false;
};

File: 653-two-sum-iv-input-is-a-bst/653-two-sum-iv-input-is-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        cache = set()
        level = [root]
        
        while level:
            for node in level:
                if node.val in cache:
                    return True
                cache.add(k - node.val)
            level = [child for node in level for child in (node.left, node.right) if child is not None]
            
        return False